test_linalg: triangular_solve - make well_conditioned well conditioned

nikitaved opened the pull request:
`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105919

nikitaved committed:
test_linalg: triangular_solve - make well_conditioned well conditioned

[ghstack-poisoned]

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` did not guaranteed the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

lezcano commented on test/test_linalg.py:
```diff
@@ -3991,7 +3991,8 @@ def _gen_shape_inputs_linalg_triangular_solve(self, shape, dtype, device, well_c
             if well_conditioned:
-                PLU = torch.linalg.lu(make_randn(*size_a))
+                U, _, Vh = torch.linalg.svd(make_randn(*size_a), full_matrices=False)
+                PLU = torch.linalg.lu(U @ Vh)
```
Probably better to simply use
```python
make_fullrank = make_fullrank_matrices_with_distinct_singular_values
```
You can set this outside the loop and use it here.

nikitaved replied:
The goal is to have condition number 1, not just full-rank, as it is not sufficient for these tests...

lezcano replied:
Not quite, we just want something that can be solved. Note that we're taking just below the `U` part if `not uni`, so that will almost never have a condition number of 1.

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

lezcano commented on test/test_linalg.py:
```diff
@@ -3991,7 +3991,9 @@ def _gen_shape_inputs_linalg_triangular_solve(self, shape, dtype, device, well_c
 
             if well_conditioned:
-                PLU = torch.linalg.lu(make_randn(*size_a))
+                make_fullrank = make_fullrank_matrices_with_distinct_singular_values
+                make_well_conditioned_arg = partial(make_fullrank, device=device, dtype=dtype)
```
nit. this could be outside of the loop

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

lezcano commented:
I think this is too overegineered. For example, you are inadvertently sampling the identity matrix when `uni=True`...

I think the full-rank fix is cleaner and faster. Let's just revert to that one.

nikitaved commented on test/test_linalg.py:
```diff
@@ -3991,13 +3992,19 @@ def _gen_shape_inputs_linalg_triangular_solve(self, shape, dtype, device, well_c
 
             if well_conditioned:
-                PLU = torch.linalg.lu(make_randn(*size_a))
+                X = make_randn(*size_a)
+                X = X + X.mH
+                U = torch.linalg.svd(X)[0]
```
Alternatively, one could use QR, but I decided to stick to SVD for probably a bit easier understanding...
But I can change that to improve efficiency if needed...

nikitaved commented:
@lezcano , sorry, where does Identity matrix come from when `uni=True`?

lezcano commented:
you are sampling an orthogonal martrix `U` and then computing `U @ U.mH`, which is just the identity matrix.

lezcano commented:
Let's roll back to the simpler fix

nikitaved commented:
No, it is not an identity, the identity is `U.mT @ U`, not `U @ U.mT`.

lezcano commented:
It's both as the orthogonal matrices form a group and `U^H = U^{-1}`. Same with the unitary matrices.

nikitaved commented:
Yeah, sorry, rushing too much. Let's revert indeed.

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

lezcano commented on test/test_linalg.py:
```diff
@@ -3972,6 +3972,7 @@ def check(*args, regex, exception=RuntimeError):
     def _gen_shape_inputs_linalg_triangular_solve(self, shape, dtype, device, well_conditioned=False):
         make_arg = partial(make_tensor, dtype=dtype, device=device)
         make_randn = partial(torch.randn, dtype=dtype, device=device)
+        make_rand = partial(torch.rand, dtype=dtype, device=device)
```
leftover?
Also ,you can lift the `make_wekk_conditioned_arg` to here.

nikitaved committed:
Update on "test_linalg: triangular_solve - make well_conditioned well conditioned"

`well_contioned=True` does not guarantee that the samples for `triangular_solve` are actually well-conditioned. This PR fixes that. This issues was discovered in https://github.com/pytorch/pytorch/pull/104425.




[ghstack-poisoned]

lezcano reviewed (approved):
Thank you!

lezcano commented:
@pytorchbot merge -i
